[iOS] Adopt API to create visibility propagation view without CA context

pvollan opened the pull request:
#### 4e208d4fe7bef5e4570932fbc00b1f08a662465f
<pre>
[iOS] Adopt API to create visibility propagation view without CA context
<a href="https://bugs.webkit.org/show_bug.cgi?id=249454">https://bugs.webkit.org/show_bug.cgi?id=249454</a>
rdar://103436616

Reviewed by Geoffrey Garen.

Adopt API to create visibility propagation view in the UI process without remote CA context from the WebContent process.
The RunningBoard assertions for the WebContent process must be created with the same environment identifier as the
visibility propagation view.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebKit/Platform/spi/ios/RunningBoardServicesSPI.h:
* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
(WebKit::ProcessAssertion::ProcessAssertion):
(WebKit::ProcessAndUIAssertion::ProcessAndUIAssertion):
* Source/WebKit/UIProcess/ProcessAssertion.cpp:
(WebKit::ProcessAssertion::ProcessAssertion):
(WebKit::ProcessAndUIAssertion::ProcessAndUIAssertion):
* Source/WebKit/UIProcess/ProcessAssertion.h:
(WebKit::ProcessAssertion::create):
* Source/WebKit/UIProcess/ProcessThrottler.cpp:
(WebKit::ProcessThrottler::setThrottleState):
* Source/WebKit/UIProcess/ProcessThrottlerClient.h:
(WebKit::ProcessThrottlerClient::environmentIdentifier const):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::environmentIdentifier const):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _setupVisibilityPropagationViewForWebProcess]):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:

Canonical link: <a href="https://commits.webkit.org/260124@main">https://commits.webkit.org/260124@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7ad29f4ddfa63b2be5ddce7575839246a383a271

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106949 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15985 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39743 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116128 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115669 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17468 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7183 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99134 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112714 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13222 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96244 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40830 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95114 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27886 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/82579 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/96455 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9129 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29238 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/95902 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/7082 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9708 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6238 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30726 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15296 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48797 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/104679 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6998 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11238 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25934 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -1614,6 +1614,14 @@ void WebProcessProxy::didChangeThrottleState(ProcessThrottleState type)
+    builder.append(processIdentifier());
+    return builder.toString();
+}
+
```
Can we retain this string as a data member instead of building it each time?

pvollan replied:
Fixed!

geoffreygaren commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -232,12 +236,22 @@ - (void)_setupVisibilityPropagationViewForWebProcess
     // Propagate the view's visibility state to the WebContent process so that it is marked as "Foreground Running" when necessary.
+#if HAVE(NON_HOSTING_VISIBILITY_PROPAGATION_VIEW)
+    auto environmentIdentifier = _page->process().environmentIdentifier();
+    auto viewClass = NSClassFromString(@"_UINonHostingVisibilityPropagationView");
```
Do we need to do NSClassFromString if we're also doing a build-time version check? What's the case where we would build this code but not have the API at runtime?

pvollan replied:
> Do we need to do NSClassFromString if we're also doing a build-time version check? What's the case where we would build this code but not have the API at runtime?

This was done to avoid compile errors in case one did not have the latest SDK, but is no longer needed. Fixed in latest patch.

geoffreygaren commented on Source/WebKit/Platform/spi/ios/RunningBoardServicesSPI.h:
```diff
@@ -52,6 +52,7 @@ NS_ASSUME_NONNULL_BEGIN
 
 @interface RBSTarget : NSObject
 + (RBSTarget *)targetWithPid:(pid_t)pid;
++ (RBSTarget *)targetWithPid:(pid_t)pid environmentIdentifier:(NSString *)environment;
```
Should we ifdef this behind HAVE_NON_HOSTING_VISIBILITY_PROPAGATION_VIEW?

pvollan replied:
I think this method has been available for a while, and is not directly tied to HAVE_NON_HOSTING_VISIBILITY_PROPAGATION_VIEW, so this may not be needed.

geoffreygaren commented on Source/WebKit/Platform/spi/ios/UIKitSPI.h:
```diff
@@ -1557,6 +1557,13 @@ typedef NS_ENUM(NSUInteger, _UIScrollDeviceCategory) {
+@interface _UINonHostingVisibilityPropagationView : _UIVisibilityPropagationView
+- (instancetype)initWithFrame:(CGRect)frame pid:(pid_t)pid environmentIdentifier:(NSString *)environmentIdentifier;
+@end
+
```
Should we ifdef this behind HAVE_NON_HOSTING_VISIBILITY_PROPAGATION_VIEW?

pvollan replied:
> Should we ifdef this behind HAVE_NON_HOSTING_VISIBILITY_PROPAGATION_VIEW?

That is a good point. Fixed in latest patch.

Thanks for reviewing!

geoffreygaren reviewed (approved):
r=me

pvollan committed:
[iOS] Adopt API to create visibility propagation view without CA context
https://bugs.webkit.org/show_bug.cgi?id=249454
rdar://103436616

Reviewed by Geoffrey Garen.

Adopt API to create visibility propagation view in the UI process without remote CA context from the WebContent process.
The RunningBoard assertions for the WebContent process must be created with the same environment identifier as the
visibility propagation view.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebKit/Platform/spi/ios/RunningBoardServicesSPI.h:
* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
(WebKit::ProcessAssertion::ProcessAssertion):
(WebKit::ProcessAndUIAssertion::ProcessAndUIAssertion):
* Source/WebKit/UIProcess/ProcessAssertion.cpp:
(WebKit::ProcessAssertion::ProcessAssertion):
(WebKit::ProcessAndUIAssertion::ProcessAndUIAssertion):
* Source/WebKit/UIProcess/ProcessAssertion.h:
(WebKit::ProcessAssertion::create):
* Source/WebKit/UIProcess/ProcessThrottler.cpp:
(WebKit::ProcessThrottler::setThrottleState):
* Source/WebKit/UIProcess/ProcessThrottlerClient.h:
(WebKit::ProcessThrottlerClient::environmentIdentifier const):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::environmentIdentifier const):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _setupVisibilityPropagationViewForWebProcess]):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:

Canonical link: https://commits.webkit.org/260124@main
